### ISSUE TITLE
Optional processor based on content-type header

### ISSUE THREAD

**Issue by simonmcl** [onevcat/Kingfisher#2085]
#### What

I have implemented a SVGImageProcessor based on the discussion in this thread: https://github.com/onevcat/Kingfisher/issues/1225 . However I am fetching third party content I don't own, from various servers, and in some situations there is no file extension in the URL for me to be able to decide ahead of time when to apply the processor. I'm looking for something in the library that would allow me to read the response's content-type header to check if its svg and then add this processor if thats the case. This doesn't appear to be possible currently (unless i'm mistaken). 

Is this functionality something you would consider adding? and in the short term do you have any guidance on how I could add it now, via subclassing, overrides, delegation etc. I don't know the library well and i'm struggling to find an easy place to add such logic

**Comment by simonmcl** [comment]
I seem to have a workaround, but I don't know what impact it will have on trying to retrieve / check if images are cached, as doing so requires passing in the processor identifier. Doing it this way, I have no way of knowing what the processor should be

Current solution:

- Fork the library
- Make various functions open instead of private, make structs public, add public init's etc
- Create an `ImageDownloader` subclass
- Override `startDownloadTask(...`
- Check the network response, if the content-type header contains "svg", then find and swap the default image processor for an svg one

```Swift

public class ContentTypeCheckingImageDownloader: ImageDownloader {
	
	public override func startDownloadTask(context: DownloadingContext, callback: SessionDataTask.TaskCallback) -> DownloadTask {
		
		...
		...
		...
			
			switch result {
					
				// Download finished. Now process the data to an image.
				case .success(let (data, response)):
					
					let processor = ImageDataProcessor(data: data, callbacks: callbacks, processingQueue: context.options.processingQueue)
					
					// If we are downloading an SVG, replace DefaultImageProcessor with SVGImgProcessor
					if let httpResponse = response as? HTTPURLResponse, let contentType = httpResponse.value(forHTTPHeaderField: "Content-Type"), contentType.contains("svg") {
						for (index, callback) in processor.callbacks.enumerated() {
							if callback.options.processor.identifier == DefaultImageProcessor.default.identifier {
								processor.callbacks[index].options.processor = SVGImgProcessor()
							}
						}
					}
					
		...
		...
		...
	}
}
```

**Comment by vanniktech** [comment]
I have the exact same requirement. I don't know what kind of image it is. My use case is an RSS Reader where the feed manages the images.

I also would have thought that `appendProcessor` would do the trick, by basically going through each processor (first the built in ones) and looking whether it can decode the image and then delegating to the appended ones; but it does not. It seems like when using `appendProcessor` it will only try very first processor. 

`setProcessor` does not seem to work either because then I loose support for the built in formats that are supported.

And neither is there the possibility to get the content type from the signature `public func process(item: ImageProcessItem, options: KingfisherParsedOptionsInfo) -> KFCrossPlatformImage? {` from which I could then try to delegate to the built in processors (if they are public).

So how should we move forward?

**Comment by vanniktech** [comment]
Obviously, the best solution would be if Kingfisher were to support SVGs by default. There's quite some demand: https://github.com/onevcat/Kingfisher/issues/1225#issuecomment-1122412772

**Comment by teameh** [comment]
Any progress on this? Would you accept a PR with native support for SVG's?
